Add DDPGfD

MrSyee opened the pull request:
code : [Colab](https://colab.research.google.com/github/MrSyee/pg-is-all-you-need/blob/feature%2Fadd-ddpgfd/06.DDPGfD.ipynb)

I add DDPGfD but it doesn't train well. I'll do debugging soon.
Please review about description and test to get the demo.pkl on colab.

And I change readme about file number. I'll change real file name on this PR.

In addition, I wonder your opinion about demo making codes named ~_for_demo.
Should those codes be removed? Or should leave?

Commit:
Add contents on README.md

Commit:
Add DDPGfd

Commit:
Add code for colab on DDPGfD

Curt-Park commented:
1. Typo: The probability of sampling a particular _transitioniis_
=> transitions

2. This one item will be better to separate:
`𝛿𝑖  is the last TD error calculated for this transition, the second term represents the loss applied to the actor.`
=>  𝛿𝑖  is the last TD error calculated for this transition.
=> the second term represents the loss applied to the actor.

3. stabilizethe final learning performance.
=> stabilize the final learning performance.

4. 
- Remove unnecessary new lines.
- Conjunctive adverb rather than conjunction.

`When the length of n_step_buffer becomes equal to N, it eventually stores the N-step transition, which is calculated by get_n_step_info method. (reference util.py)

And there are additional implementations for saving loaded demos.

(Please see 03.DDPG.ipynb for detailed description of the basic replay buffer.)`
=>
When the length of n_step_buffer becomes equal to N, it eventually stores the N-step transition, which is calculated by get_n_step_info method (reference util.py). Furthermore, there are more implementations for saving loaded demos. (Please see 03.DDPG.ipynb for the description of the basic replay buffer.)

5. I think creating `util.py` for two additional functions is not a good idea for code readability on Colab or Jupyter. How about including each function at the proper position? For example, `get_n_step_info` can be located above `ReplayBuffer` class, and `get_n_step_info_from_demo` can be defined before the agent class. 

6. `ReplayBuffer` doesn't have docstrings for each method.

7. The following line is too long.
`            indices = np.random.choice(len(self), size=self.batch_size, replace=False)`

8. In the description of Network and ActionNormalizer, _Italic_ is better than **Bold**.

9. In the table for the summary of Agent class.
`plot` => `_plot`

10. I suggest a more simple implementation of the step function.

```
def step(self, action: np.ndarray) -> Tuple[np.ndarray, np.float64, bool]:
        """Take an action and return the response of the env."""
        next_state, reward, done, _ = self.env.step(action)
        
        if not self.is_test:
            self.transition += [reward, next_state, done]
            
            # N-step transition
            if self.use_n_step:
                self.transition = self.memory_n.store(*self.transition)

            # add a single step transition
            if self.transition:
                self.memory.store(*self.transition)
    
        return next_state, reward, done
```

11. the following lines are too long.
`weights = torch.FloatTensor(samples["weights"].reshape(-1, 1)).to(device)`
`critic_loss_n_element_wise = self._get_critic_loss(samples_n, n_gamma)`
`critic_loss_element_wise += critic_loss_n_element_wise * self.lambda1`

12. In sample_batch method in PER class, is epsilon_d added for every transition, not only demo transition?
`epsilon_d = np.ones(self.batch_size) * self.epsilon_d`
...
`new_priorities += epsilon_d`

Curt-Park commented:
In terms of the demo file:

1. we can retain a feature branch which includes the demo generating code. On master branch, we can just place the demo file.
2. If you just place the demo file on the top directory, you don't have to use `mv` and `mkdir` commands.
3. We should more investigate the demo file to check up why the performance is not so great. How did you gather the demo transitions? and how many transitions did you gather? 
I doubt the issue possibly comes from high correlations from the data.

MrSyee commented:
> In terms of the demo file:
> 
> 1. we can retain a feature branch which includes the demo generating code. On master branch, we can just place the demo file.
> 2. If you just place the demo file on the top directory, you don't have to use `mv` and `mkdir` commands.
> 3. We should more investigate the demo file to check up why the performance is not so great. How did you gather the demo transitions? and how many transitions did you gather?
>    I doubt the issue possibly comes from high correlations from the data.

1. I agree.
2. I agree.
3. It gathers the demo transitions for 2000 steps (10 episodes). It saves a list with demo transitions that the trained-well network(A2C or DDPG, performances of both are similar) makes for 10 episodes. (see the bottom of [*ddpg_for_demo.ipynb*](https://colab.research.google.com/github/MrSyee/pg-is-all-you-need/blob/feature%2Fadd-ddpgfd/DDPG_for_demo.ipynb))

I'll more investigate the parameters and demo files.

Thank you for your review.

MrSyee commented:
@Curt-Park How do you check line length? I didn't know that line length is over on jupyter notebook.

Curt-Park commented:
@MrSyee 

1. Colab states the proper line length.
<img width="1061" alt="Screen Shot 2019-09-14 at 4 11 07 PM" src="https://user-images.githubusercontent.com/14961526/64904847-ca964000-d70a-11e9-9b66-6ad46fa3d728.png">

2. please check the logic regarding epsilon_d usage.
I described it in [my 12th comment](https://github.com/MrSyee/pg-is-all-you-need/pull/6#issuecomment-531133593).

MrSyee commented:
epsilon_d would be my mistake. I'll check and modify it.

MrSyee commented:
@Curt-Park How about changing line limit 80 to 88? We don't use `black`, but I think 88 line limit would be better. In colab, you can change the line limit on `Tools - Preference - Editor` tab.

MrSyee committed:
Fix typo

Curt-Park commented:
@MrSyee can we save the configuration and deploy?

MrSyee committed:
Fix wrong epsilon_d on PER

Commit:
Remove utils.py

MrSyee commented:
@Curt-Park I've modified about your review.
1, 2, 3, 4, 5, 8, 9, 10. Fix them.
6. The other implements (DDPG, SAC) also don't have detail documents. I think classes except for agent class need not have detail docstrings like the agent class.
7, 11. Mentioned above, the line limit should be changed. ([colab](https://colab.research.google.com/github/MrSyee/pg-is-all-you-need/blob/feature%2Fadd-ddpgfd/06.DDPGfD.ipynb))
12. Fix it.
```
epsilon_d = np.array([self.epsilon_d if i < self.demo_size else 0.0 for i in indices])
```

Unfortunately, It hasn't trained yet. :(
![image](https://user-images.githubusercontent.com/17582508/64918817-7197dc00-d7de-11e9-9cd3-9ef522320cee.png)

I'll add pre-train.

Curt-Park commented:
@MrSyee 
6. linting setting should be consistent. Why do you think it doesn't need to have docstring for methods? **Exception** always needs to have any proper reason.
7, 11. I asked the collab configration can be deployed. On your link, the line limit is still 80.

MrSyee commented:
@Curt-Park 
> @MrSyee
> 6. linting setting should be consistent. Why do you think it doesn't need to have docstring for methods? **Exception** always needs to have any proper reason.
- I think it's too long to write detail docstring about all class. I want that it has good readability. Also, your repo (rainbow-is-all-you-need) doesn't have all docstring. I agree with the style of your repo and implement the same.
> 7, 11. I asked the collab configration can be deployed. On your link, the line limit is still 80.
- Well. Depends on me, the line limit is 88 on this link. It can't be deployed. I consider this problem.

Curt-Park commented:
@MrSyee 
6. To be honest, I just missed docstrings on ReplayBuffer in DQN2Rainbow, and there was no reason to skip it. Anyways, this project lead is you. You can decide it.
7, 11. If it can't be deployed, we should comply with the default settings in my opinion.

MrSyee commented:
@Curt-Park 
> 6. To be honest, I just missed docstrings on ReplayBuffer in DQN2Rainbow, and there was no reason to skip it. Anyways, this project lead is you. You can decide it.
- I misunderstand your review. I mean that I'll add one line docstring for method, but it doesn't need attribute docstrings for all classes. I'll add a missing docstring.
> 7, 11. If it can't be deployed, we should comply with the default settings in my opinion.
- I agree with your opinion.

Commit:
Add pre-train, add docstring for method

MrSyee commented:
@Curt-Park [colab](https://colab.research.google.com/github/MrSyee/pg-is-all-you-need/blob/feature%2Fadd-ddpgfd/06.DDPGfD.ipynb)
I've modified code about the above discussion (line limit, docstring for methods).
In addition, I've added a pre-train method on agent class. It makes a good result. If you don't find another problem, it can be merged.

Commit:
Remove code for making demonstration

Commit:
Add description of DDPG

Curt-Park commented:
1. it misses descriptions of some attributes in DDPGAgent docstrings.
2. _pretrain method should states the return type: Tuple[List[torch.Tensor], List[torch.Tensor]]
3. in _pretrain method, i_step is unnecessary variable. You can just write as follows:
```
for _ in range(self.pretrain_step):
        actor_loss, critic_loss, _ = self.update_model()
        actor_losses.append(actor_loss.data)  # .data!
        critic_losses.append(critic_loss.data)  # .data!
```
4. we can simply state
```
        actor_losses, critic_losses, n_demo_list, scores = [], [], [], []
```
5. we don't have to print out the pretraining result.
```
        self._plot(
            self.total_step, 
            scores, 
            actor_losses, 
            critic_losses,
            n_demo_list,
        )
```
Instead, you can just modify the statement in the training loop:
```
for self.total_step in range(num_frames + 1):  # starts from 0
```
6. How about align the order of public methods and protected methods of DDPGAgent class both in the code and the table of the description? (public first, protected latter)
7. The result is not so great in my execution.
![untitled](https://user-images.githubusercontent.com/14961526/64921510-2e019a00-d7ff-11e9-8996-c40f99808279.png)

MrSyee committed:
Change docstring and etc

Curt-Park commented on 06.DDPGfD.ipynb:
```diff
@@ -808,15 +813,15 @@
     "        critic_losses = []\n",
     "        print(\"Pre-Train %d step.\" % self.pretrain_step)\n",
-    "        for i_step in range(1, self.pretrain_step + 1):\n",
+    "        for _ in range(1, self.pretrain_step + 1):\n",
```
why not `range(self.pretrain_step)`?

MrSyee commented:
1, 2, 3, 4, 6. I've modified them.
5. I agree, but I think I don't have to change the training loop.
7. It is trained depends on random seed. It would be wrong. I'll check again.

Curt-Park commented:
@MrSyee 5. Simple is the best. It doesn't look concise.

it's the same as we don't state `return` in `else` statement in the following situation:

```
def return_something():
    if condition:
        return one
    return another  # we don't usually state else: return another
```

Come and join CPPS.

Commit:
Merge branch 'feature/add-ddpgfd' of https://github.com/MrSyee/pg-is-all-you-need into feature/add-ddpgfd

Commit:
Fix the wrong input on 1-step buffer

Squashed commit of the following:

commit 86e814b788699665bae770cc1fc6cc766dc94377
Author: khkim <kh.kim@medipixel.io>
Date:   Wed Sep 18 10:30:26 2019 +0900

    Remove ddpgfd for testing

commit 8d65ba1df80ddd392d557d11963b609dfe898972
Author: khkim <kh.kim@medipixel.io>
Date:   Wed Sep 18 09:39:20 2019 +0900

    Move random seed cell to bottom

commit fefbfc683879fd2a6785786c4ff06491c4acb2f4
Author: khkim <kh.kim@medipixel.io>
Date:   Wed Sep 18 00:01:26 2019 +0900

    Test ddpgfd2

commit e8d3a1cc84f5664ee93e09073bca830ea8ae8584
Author: khkim <kh.kim@medipixel.io>
Date:   Tue Sep 17 23:46:37 2019 +0900

    Test ddpgfd

commit b4ac5cef3a3f5b8b5c9982b8924d1780937ced6a
Author: khkim <kh.kim@medipixel.io>
Date:   Tue Sep 17 23:18:33 2019 +0900

    Add fd

commit 33fc0dd8cf310a0ea6ac8df47d30f1907184152e
Author: khkim <kh.kim@medipixel.io>
Date:   Tue Sep 17 22:15:49 2019 +0900

    Add n-step ddpg PER

commit c628f8ba6189448bf5bc12d73e8032698409afcb
Author: khkim <kh.kim@medipixel.io>
Date:   Tue Sep 17 21:49:05 2019 +0900

    Verify ddpg PER

commit d85062e789065bf1c9a7de20c7bc833978bfcc21
Author: khkim <kh.kim@medipixel.io>
Date:   Tue Sep 17 20:38:51 2019 +0900

    Remove n-step and demo

MrSyee commented:
I fix a critical bug. I'm typing the wrong value as alpha to the place where I'm typing the gamma.
```
        self.memory = PrioritizedReplayBuffer(
            obs_dim, memory_size, batch_size, gamma, alpha, demo=demos_1_step
        )
```
And I modify variable names and docstring. If you see the diff, please type `nbdiff-web HEAD~1` after pull.

I check the success of the test in every case. However, I wonder that the score graph for the training phase looks not good while test score is good.
Please check [colab](https://colab.research.google.com/github/MrSyee/pg-is-all-you-need/blob/feature%2Fadd-ddpgfd/06.DDPGfD.ipynb).

mclearning2 reviewed (approved):
I tested it 5 times. The graph doesn't look so good. However, It performs great!!
![image](https://user-images.githubusercontent.com/43226417/65290859-d4e18f80-db8b-11e9-926b-dde4f7cee5e5.png)

Curt-Park commented:
@MrSyee 
I tuned the noise parameters theta and sigma and got better results.
We should carefully tune the parameters considering the magnitude of the action and disturbance in the environment. In this environment, we normalize the action range into [-1, 1] and there is no disturbance. Due to that, if we set sigma a large value even greater than theta, the agent suffers from disturbance and cannot expect where the noise goes.
I recommend the parameter settings as follows:
```
ou_noise_theta = 1.0
ou_noise_sigma = 0.1
```

This is the training result. I couldn't run the test because the session terminated by the connection issue. I guess it would be great.
![beutibul_res](https://user-images.githubusercontent.com/14961526/65382466-fc2d8d80-dd40-11e9-867b-c9cbf0410775.png)

MrSyee committed:
Change ou noise parameters

MrSyee commented:
Great! Thank you. I'll try to fix this parameter on DDPG.

MrSyee commented:
DDPG also makes a good graph. I push this change.
![image](https://user-images.githubusercontent.com/17582508/65390680-61fb3300-dd9c-11e9-940f-e6786c7531f8.png)

MrSyee committed:
Change ou noise parameter on DDPG

Curt-Park commented:
@MrSyee Don't forget to change the following lines:
```
if IN_COLAB and not os.path.exists("segment_tree.py"):
    # download segment tree module
    # TODO: Modify link after merging master
    !wget https://raw.githubusercontent.com/mrsyee/pg-is-all-you-need/feature/add-ddpgfd/segment_tree.py
    # download demo.pkl
    !wget https://raw.githubusercontent.com/mrsyee/pg-is-all-you-need/feature/add-ddpgfd/demo.pkl
```

MrSyee commented:
Thank you for your remind. I'll change it on BC PR.
